crates/steel-repl/src/highlight.rs

```diff
@@ -44,11 +39,36 @@ impl RustylineHelper {
 
 impl Validator for RustylineHelper {
     fn validate(&self, ctx: &mut ValidationContext) -> rustyline::Result<ValidationResult> {
-        self.validator.validate(ctx)
-    }
+        use steel_parser::tokens::TokenType;
+
+        let token_stream = TokenStream::new(ctx.input(), true, None);
+
+        let mut balance = 0;
+
+        let mut unfinished_string = false;
+
+        for token in token_stream {
+            unfinished_string = match token.ty {
+                TokenType::Error => token.source.starts_with("\""),
+                _ => false,
+            };
 
-    fn validate_while_typing(&self) -> bool {
-        self.validator.validate_while_typing()
+            match token.ty {
+                TokenType::OpenParen(_) => {
+                    balance += 1;
+                }
+                TokenType::CloseParen(_) => {
+                    balance -= 1;
+                }
+                _ => {}
+            }
+        }
+
+        if balance > 0 || unfinished_string {
+            Ok(ValidationResult::Incomplete)
+        } else {
+            Ok(ValidationResult::Valid(None))
+        }
```

Comment:
I'm not a big fan of returning `ValidationResult::Invalid`. The message appears right next to the input and it's not very user-friendly IMO:
```
λ > (aadf]Mismatched brackets: '(' is not properly closed
```

So the case of `balance < 0` or mismatched parentheses is left to the parser to throw a much more precise error.